### ISSUE TITLE
"Dirty" variables incorrectly reset with a multiple camera setup causing drawing issues on Sprites

### ISSUE THREAD

**Issue by wildrabbit** [cocos2d/cocos2d-x#16100]
On our project we have a scene with two cameras: the default one, with 2D nodes (mostly sprites), and a new one we create ourselves to render a 3D scene. Some of the nodes rendered using the default camera begin initially offscreen, and at a later point we scroll them to enter the visible area. However, they're not showing.

After debugging for quite a while to rule out possible problems on our client code related to clipping or incorrect positioning, we found evidence pointing to the two cameras as the likely suspect. I started digging a bit deeper and spotted a potential bug on the visiting/drawing logic.

The method Node::processParentFlags() resets the variables _transformUpdated and    _contentSizeDirty, and that happens as soon as we visit the node with the first of the scene cameras, regardless that the node's camera mask determines that it won't be rendered by that camera. Interestingly, there's a commented check on processParentFlags() that checks if the node will be visited by the current camera and returns early if it won't.

```
uint32_t Node::processParentFlags(const Mat4& parentTransform, uint32_t parentFlags)
{
    if(_usingNormalizedPosition)
    {
        CCASSERT(_parent, "setNormalizedPosition() doesn't work with orphan nodes");
        if ((parentFlags & FLAGS_CONTENT_SIZE_DIRTY) || _normalizedPositionDirty)
        {
            auto& s = _parent->getContentSize();
            _position.x = _normalizedPosition.x * s.width;
            _position.y = _normalizedPosition.y * s.height;
            _transformUpdated = _transformDirty = _inverseDirty = true;
            _normalizedPositionDirty = false;
        }
    }

    //remove this two line given that isVisitableByVisitingCamera should not affect the calculation of transform given that we are visiting scene
    //without involving view and projection matrix.

//    if (!isVisitableByVisitingCamera())
//        return parentFlags;

    uint32_t flags = parentFlags;
    flags |= (_transformUpdated ? FLAGS_TRANSFORM_DIRTY : 0);
    flags |= (_contentSizeDirty ? FLAGS_CONTENT_SIZE_DIRTY : 0);


    if(flags & FLAGS_DIRTY_MASK)
        _modelViewTransform = this->transform(parentTransform);

    _transformUpdated = false;
    _contentSizeDirty = false;

    return flags;
}
```

Then let's change to a Sprite's draw() method.  

```
    if (visitingCamera == defaultCamera) {
        _insideBounds = ((flags & FLAGS_TRANSFORM_DIRTY)|| visitingCamera->isViewProjectionUpdated()) ? renderer->checkVisibility(transform, _contentSize) : _insideBounds;
    }
```

When a sprite node beings its "execution" offscreen, the variable _insideBounds is correctly set to false. However, because of the issue with the flags being reset by the first camera that visits the node, it never gets to enter the checkVisibility() method again, so it remains false and doesn't get drawn again.

Altering the default camera depth gets those pesky nodes drawn, but then we get other nodes hidden because of the rendering order, so that option doesn't work for us. Another way we might fix the problem on our side could be by defining another camera and render those sprites using it.  Even uncommenting those two lines on processParentFlags seems to fix the issue, but I agree on the reasoning behind it in the comments.

Either way, the behaviour itself feels wrong to me in principle, those two variables on processParentFlags shouldn't be reset until all the cameras have visited them and it's safe to do so. 

//-------------------

Engine version: 3.9, but from what I can see the source code hasn't changed much since then on the v3 branch.

Steps to reproduce: 
1: Create a scene with an additional camera. Set the new camera's flag to a non-default value, and leave both its depth and the default camera's to use their default values (-1 and 0 respectively)
2: Add some sprites to the scene, set their masks to be rendered by either of the cameras, and make sure that at least one of the sprites rendered by the default camera is initially placed off-screen.
3: Use an action or move the off-screen sprite(s) manually until they enter the scene bounds.

**Comment by minggo** [comment]
@ricardoquesada could you please take a look? thanks.

**Comment by ricardoquesada** [comment]
@wildrabbit good finding!

Yes, I guess commenting out those two lines in `Node::processParentFlags()` should fix the problem... 
I'll create a test case to prove it. thanks.

**Comment by ricardoquesada** [comment]
@wildrabbit I created a test case, but I couldn't reproduce the bug... in theory, I understand the bug, but for some reason I cannot reproduce it.
Could post the code the reproduce it? or take a look at my test case and see what's wrong ? Thanks.
Here is my test case:
https://github.com/cocos2d/cocos2d-x/pull/16122

**Comment by wildrabbit** [comment]
I've tried the test at home with a 3.10 installation and I can't reproduce it either. The sprite on the default camera never gets to change the value of "insideBounds" on the draw() method, which was our problem. 

I'll take a look at work tomorrow to see exactly at which point our sprites got that variable set to false (perhaps because of the cameras viewProjectionUpdated member, but I'm not sure at the moment) so we can try to reproduce it on the test. 

Thanks!

**Comment by wildrabbit** [comment]
I've seen what is happening in the test case :D
Funny enough, we're actually reproducing the bug in the test in a way that prevents us from noticing it.

Debugging our project right now, I've noticed that at the moment the "_insideBounds" variable was set to false there was only the default camera on the scene instead of two cameras. Because of this, the node variables controlling the flags don't get reset for the sprites' first draw() call with the default camera, and the initial visibility check correctly sets _insideBounds to false. Our second camera is then added to the scene a bit later, but before the node gets to enter the visible bounds, and it is then that the bug manifests in all its glory, as the second camera starts invalidating the dirty variables and _insideBounds never goes back to true after that.

In the test case, however, we had both cameras ready by the time the sprites are first drawn. This means that the flags are incorrectly reset for the default camera, and the sprite is never really checked for visibility. We just can't spot the wrong behaviour because when it actually enters the scene bounds, _insideBounds has never changed from its initialization value, true. 

If in the test case we delay the creation of the custom camera for a short while, just the necessary time to get the first sprite::draw() call before the moving node enters the scene bounds, then we can see the problem. 

This is what I changed on void Issue16100Test::onEnter() to reproduce it AND see it. 
A 0.4s delay worked fine for me on my machine, I also tried 1.4s for a longer duration and the sprite became visible before the second camera was added.

```
    auto delay = DelayTime::create(0.4f);
    auto f = CallFunc::create([this, s]() 
    {
        auto camera = Camera::createOrthographic(s.width * 2, s.height * 2, -1024, 1024);
        camera->setCameraFlag(CameraFlag::USER1);
        addChild(camera); 
    });

    this->runAction(Sequence::createWithTwoActions(delay, f));
```

By the way, commenting out the two lines in processParentFlags() does seem to fix it indeed.

**Comment by ricardoquesada** [comment]
Thanks! I'll update everything on Tuesday.

**Comment by ricardoquesada** [comment]
@wildrabbit thanks for the info. However I couldn't reproduce it with the delay. This is what I noticed:

If I use a delay between 0 and 1 seconds, then I don't notice the bug.
If I use a delay bigger than 1s, then the second sprite is not visible when it enters the screen, which makes sense since its camera was not created yet. Once it is created, the sprite is visible again.

Am I doing something wrong? I've just updated the Pull Request with your suggestions. Thanks!

**Comment by wildrabbit** [comment]
Thanks, Ricardo!

I'll take a look tonight again or tomorrow, will let you know if I can find something.

**Comment by ricardoquesada** [comment]
@wildrabbit thanks. I just want to have a test case that triggers the bug before patching it. Thanks again!

**Comment by wildrabbit** [comment]
I think I've spotted the reason why we stop reproducing the bug on later versions. However, now I'm unsure about what would be the right way to proceed after this.

As I mentioned, I had been testing versions equal to, or closer to 3.9 , which is what we use at work (as there are multiple projects relying on the library and our own layer on top of cocos we can't just upgrade to the latest version immediately). I've been trying with different commits and the inflection point seems to be between these two (well, one of them squashes a branch's commits): 

[https://github.com/cocos2d/cocos2d-x/commit/a348cbda8b9e9c493eff445dcdf578449d18f70e#diff-8f0b433b0f127dff182489dc4486a85f] Buggy behaviour can be reproduced and seen.

and 

[https://github.com/cocos2d/cocos2d-x/commit/180c2f1e9a6257d9d161969fb169039b1ee0b5a1#diff-8f0b433b0f127dff182489dc4486a85f] We stop seeing the bug.

The weird part now is that, debugging Sprite::draw(), when we render Grossini with the default camera we still see that the flags have been incorrectly reset. Why does it work, then? Because the other clause of the OR that controls the visibility checks, visitingCamera->isViewProjectionUpdated() returns true.

The reason why that happens now seems to be because of some new logic added for VR support on Scene::render(). There, a new call to Node::setAdditionalTransform() is performed on the cameras which, among other things, sets the _transformUpdated variable to true. Later, Camera::visit() will use that value to update _viewProjectionUpdated. In fact, if I remove from the scene rendering code the call to setAdditionalTransform the bug appears again, but this hack alters the main code and the test results shouldn't be valid any more, so let's forget about it.

So, to summarize: 
- Flags are still reset when they shouldn't.
- Due to the new _setAdditionalTransform_ calls on scene rendering, the camera has its _viewProjectionUpdated_ value to true, and sprites can perform the visibility checks when drawn on the default camera. even if their own "dirty" flags have been set to 0.

I don't know if with these findings it would make sense to close the issue or try to find a different way to reproduce it. On the one hand, there is still a potentially incorrect behaviour with the flags, but on the other it's much harder to devise a way to test where else it might cause problems. And besides, the newest changes might have additional side effects in return.

Anyway, thank you very much! :)

**Comment by ricardoquesada** [comment]
@wildrabbit thanks for all the findings.
I was able to reproduce the bug by removing the the `setAdditionalTransform` in `CCScene.cpp`.
Once the PR is green, I'll merge it. Thanks again!

**Comment by ricardoquesada** [comment]
fixed. thanks.

**Comment by wildrabbit** [comment]
That's brilliant, thank you! :)
